### PR TITLE
Add global filter chain configuration API part 2

### LINKED ISSUE #132: Configuration secret isn't DRY and can exceed K8s annotation/secret size limits
We utilize authservice behind a lot of different virtual services, however the way the current configuration file exists, we end up needing to create separate chains for every application, even though nearly all the contents are repeated. Here's a redacted version of the configuration file with 2 applications, containing the same values across both:

```
{
	"listen_address": "0.0.0.0",
	"listen_port": "10003",
	"log_level": "trace",
	"threads": 8,
	"chains": [
		{
			"name": "chain-1",
			"match": {
				"header": ":authority",
				"prefix": "foo.example.com"
			},
			"filters": [
				{
					"oidc": {
						"authorization_uri": "https://keycloak.example.com/auth/realms/example-realm/protocol/openid-connect/auth",
						"token_uri": "https://keycloak.example.com/auth/realms/example-realm/openid-connect/token",
						"callback_uri": "https://foo.example.com/login",
						"jwks": "MATCHES-ACROSS-CHAINS",
						"client_id": "client-id-1",
						"client_secret": "client-password-1",
						"scopes": [],
						"cookie_name_prefix": "foo-cookie-prefix",
						"id_token": {
							"preamble": "Bearer",
							"header": "Authorization"
						},
						"access_token": {
							"header": "JWT"
						},
						"logout": {
							"path": "/logout",
							"redirect_uri": "https://keycloak.example.com/auth/realms/example-realm/protocol/openid-connect/token/logout"
						}
					}
				}
			]
		},
		{
			"name": "chain-2",
			"match": {
				"header": ":authority",
				"prefix": "bar.example.com"
			},
			"filters": [
				{
					"oidc": {
						"authorization_uri": "https://keycloak.example.com/auth/realms/example-realm/protocol/openid-connect/auth",
						"token_uri": "https://keycloak.example.com/auth/realms/example-realm/openid-connect/token",
						"callback_uri": "https://bar.example.com/login",
						"jwks": "MATCHES-ACROSS-CHAINS",
						"client_id": "client-id-1",
						"client_secret": "client-password-1",
						"scopes": [],
						"cookie_name_prefix": "bar-cookie-prefix",
						"id_token": {
							"preamble": "Bearer",
							"header": "Authorization"
						},
						"access_token": {
							"header": "JWT"
						},
						"logout": {
							"path": "/logout",
							"redirect_uri": "https://keycloak.example.com/auth/realms/example-realm/protocol/openid-connect/token/logout"
						}
					}
				}
			]
		},
	]
}
```
In the above example, given we're using a shared client ID and secret across chains, the only differencing fields are:
* Chain `name`
* `match.prefix`
* ` filters.[].oidc.callback_uri`
* ` filters.[].oidc.cookie_name_prefix`

If the configuration could be templated out or dynamically generated to make the configuration DRY, short, and as simple as possible, it'd make our use-case infinitely easier.

### PR DESCRIPTION
**Pull request by Shikugawa**
Close #132

### CODE REVIEW

**Comment by Shikugawa** [comment]
cc: @ZackButcher @lizan

**Comment by ZackButcher** [comment]
Can you update https://github.com/istio-ecosystem/authservice/blob/master/docs/README.md (which looks generated, but I haven't looked into how yet), and add an example using the override somewhere so it's easy for someone to see how to use it?

File: test/filters/filter_chain_test.cc [hunk 1]
```diff
@@ -87,5 +88,32 @@ TEST(FilterChainTest, New) {
...
 }
 
+TEST(FilterChainTest, Override) {
+  auto default_config = config::oidc::OIDCConfig();
+  default_config.set_authorization_uri("https://istio.io/auth/default");
+  default_config.set_token_uri("https://istio.io/token");
+  default_config.set_jwks("default_jwk");
+
+  auto configuration =
+      std::unique_ptr<config::FilterChain>(new config::FilterChain);
```
Comment (lizan) [inline test/filters/filter_chain_test.cc:98]: nit: make_unique

**Comment by cfryanr** [comment]
> Can you update https://github.com/istio-ecosystem/authservice/blob/master/docs/README.md (which looks generated, but I haven't looked into how yet), and add an example using the override somewhere so it's easy for someone to see how to use it?

In case it helps, you can regenerate the docs with `make docs` and then commit the updated doc. See [Makefile](https://github.com/istio-ecosystem/authservice/blob/master/Makefile#L10-L13).

**Review by lizan — APPROVED** [review]

**Pull request merged**